File: Taller_1/Profile_Algorithms/Algs_java/a3.java
/*

Big O total = O(n) + O(1) = O(n)

`operacionIntensivaMemoria`
Big O = O(n), contiene un bucle que se ejecuta 'n' veces, donde 'n' es el tamaño de la lista `granLista`.
Dentro de este bucle, se realiza una operación de tiempo constante (añadir un elemento a la lista), por lo que la complejidad total es O(n).


`operacionIntensivaCPU`
Big O = O(1),  a pesar del tamaño de la entrada, siempre realiza 100 operaciones (el bucle se ejecuta 100 veces).

 */


// Importamos las clases ArrayList y List, y la clase Random de la biblioteca java.util
import java.util.ArrayList;
import java.util.List;
import java.util.Random;



public class a3 {
    // Método principal que se ejecuta al iniciar el programa.
    public static void main(String[] args) throws InterruptedException {
        // Definimos un entero 'n' con el valor de 500000.
        int n = 500000;
        // Bucle for que se ejecuta 5 veces.
        for (int i = 0; i < 5; i++) {
            // Imprimimos en consola el inicio de una operación intensiva en memoria.
            System.out.println("Pico " + (i + 1) + ": Operación intensiva en memoria");
            // Llamamos al método 'operacionIntensivaMemoria' con 'n' como argumento.
            operacionIntensivaMemoria(n);
            // Imprimimos en consola el inicio de una operación intensiva en CPU.
            System.out.println("Pico " + (i + 1) + ": Operación intensiva en CPU");
            // Llamamos al método 'operacionIntensivaCPU' con 100 como argumento.
            operacionIntensivaCPU(100);
        }
    }

    // Método 'operacionIntensivaMemoria' que toma un entero 'n' como argumento y devuelve un double.
    // Este método simula una operación intensiva en memoria.
    public static double operacionIntensivaMemoria(int n) throws InterruptedException {
        // Creamos un objeto 'rand' de la clase Random.
        Random rand = new Random();
        // Creamos una lista de Doubles.
        List<Double> granLista = new ArrayList<>();
        // Bucle for que se ejecuta 'n' veces.
        for (int i = 0; i < n; i++) {
            // Añadimos un número double aleatorio a 'granLista'.
            granLista.add(rand.nextDouble());
        }
        // Hacemos que el hilo actual se duerma durante 1000 milisegundos.
        Thread.sleep(1000);
        // Devolvemos la suma de todos los elementos en 'granLista'.
        return granLista.stream().mapToDouble(Double::doubleValue).sum();
    }

    // Método 'operacionIntensivaCPU' que toma un entero 'n' como argumento.
    // Este método simula una operación intensiva en CPU.
    public static void operacionIntensivaCPU(int n) throws InterruptedException {
        // Creamos un objeto 'rand' de la clase Random.
        Random rand = new Random();
        // Creamos un contador y lo inicializamos en 0.
        double contador = 0;
        // Bucle for que se ejecuta 'n' veces.
        for (int i = 0; i < n; i++) {
            // Añadimos un número double aleatorio al contador.
            contador += rand.nextDouble();
        }
        // Hacemos que el hilo actual se duerma durante 10 milisegundos.
        Thread.sleep(10);
    }
}

File: Taller_1/Profile_Algorithms/Algs_py/a3.py
# Importamos el módulo random para generar números aleatorios
import random
# Importamos el módulo time para trabajar con funciones relacionadas con el tiempo
import time
# Importamos el módulo memory_profiler para analizar el uso de memoria
from memory_profiler import profile

@profile
def operacion_intensiva_memoria(n):
    """
    Esta función genera una gran cantidad de uso de memoria temporalmente.

    Args:
        n (int): El número de elementos en la lista generada.

    Returns:
        float: La suma de los elementos en la lista generada.
    """
    # Generamos una lista grande con números aleatorios
    gran_lista = [random.random() for _ in range(n)]
    time.sleep(1)  # Simulamos un procesamiento
    # Devolvemos la suma de los elementos en la lista
    return sum(gran_lista)

@profile
def operacion_intensiva_cpu(n):
    """
    Esta función consume tiempo de CPU.

    Args:
        n (int): El número de iteraciones para el bucle.

    Returns:
        None
    """
    # Inicializamos un contador
    contador = 0
    # Iteramos n veces
    for _ in range(n):
        # Aumentamos el contador con un número aleatorio
        contador += random.random()
        time.sleep(0.01)  # Añade un pequeño retardo para simular procesamiento

def main():
    """
    Función principal que llama a las funciones operacion_intensiva_memoria y operacion_intensiva_cpu.

    Returns:
        None
    """
    # Definimos el número de elementos para la lista grande
    n = 500000
    # Iteramos 5 veces
    for i in range(5):
        # Imprimimos el pico de operación intensiva en memoria
        print(f"Pico {i+1}: Operación intensiva en memoria")
        # Llamamos a la función operacion_intensiva_memoria
        operacion_intensiva_memoria(n)
        # Imprimimos el pico de operación intensiva en CPU
        print(f"Pico {i+1}: Operación intensiva en CPU")
        # Llamamos a la función operacion_intensiva_cpu
        operacion_intensiva_cpu(100)

# Verificamos si este script se está ejecutando como el principal
if __name__ == "__main__":
    # Llamamos a la función main
    main()